inngest-core/src/main/kotlin/com/inngest/Step.kt

```diff
@@ -125,4 +134,37 @@ class Step(val state: State) {
             throw StepInterruptSendEventException(id, hashedId, response!!.ids)
         }
     }
+
+    /**
+     * Waits for an event with the name provided in `waitEvent`, optionally check for a condition
+     * specified in `ifExpression`
+     *
+     * @param id Unique step id for memoization.
+     * @param waitEvent The name of the event we want the function to wait for
+     * @param timeout The amount of time to wait to receive an event. A time string compatible with https://www.npmjs.com/package/ms
+     * @param ifExpression An expression on which to conditionally match the original event trigger (`event`) and the wait event (`async`).
+     *        Expressions are defined using the Common Expression Language (CEL) with the events accessible using dot-notation.
+     *
+     */
+    fun waitForEvent(
+        id: String,
+        waitEvent: String,
+        timeout: String,
+        ifExpression: String?,
+        // TODO use better types for timeout and ifExpression that serialize to the relevant strings we send to the inngest server, instead of using raw strings
+        // TODO support `match` which is a convenience for checking the same expression in `event` and `async`. Also make it a mutually exclusive argument with
+        // ifExpression, possibly with a sealed class?
+    ): Any? {
+        val hashedId = state.getHashFromId(id)
+
+        try {
+            val stepResult = state.getState<Any?>(hashedId)
+            if (stepResult != null) {
+                return stepResult
+            }
+            return null // TODO should this throw an exception?
```

Comment:
It can also return `EventPayload` as well i think.
https://www.inngest.com/docs/reference/functions/step-wait-for-event#step-wait-for-event-id-options-promise-null-event-payload
https://github.com/inngest/inngest-py/blob/4547a366bdab61b6fffab04db4d6c85a7191ea98/inngest/_internal/step_lib/step_async.py#L387-L393

Reply:
adding a TODO